lib/elixir/test/elixir/module/types/expr_test.exs

```diff
@@ -193,12 +201,95 @@ defmodule Module.Types.ExprTest do
       assert {:error, _} = quoted_expr([a], case(a, do: (:foo = b -> :bar = b)))
 
       assert quoted_expr([a], case(a, do: (:foo = b -> :foo = b))) == {:ok, :dynamic}
+
+      assert quoted_expr(
+               [a],
+               case a do
+                 :foo = b -> b
+                 :bar = b -> b
+               end
+             ) == {:ok, :dynamic}
+
+      assert quoted_expr(
+               [a],
+               case a do
+                 b when is_atom(b) -> b
+                 b when is_integer(b) -> b
+               end
+             ) == {:ok, :dynamic}
+
+      assert quoted_expr(
+               [a, b],
+               case a do
+                 :foo when is_binary(b) -> b
+               end
+             ) == {:ok, :dynamic}
+
+      assert quoted_expr(
+               [a, b],
+               case a do
+                 :foo when is_binary(b) -> b <> ""
+                 :foo when is_list(b) -> b
+               end
+             ) == {:ok, :dynamic}
+
+      assert quoted_expr(
+               [a],
+               case a do
+                 :foo = b -> :foo = b
+                 :bar = b -> :bar = b
+               end
+             ) == {:ok, :dynamic}
+
+      assert quoted_expr(
+               [a],
+               case a do
+                 :foo ->
+                   b = :foo
+                   b
+
+                 :bar ->
+                   b = :bar
+                   b
+               end
+             ) == {:ok, :dynamic}
+
+      assert quoted_expr(
+               [a, b],
+               case a do
+                 :foo -> :foo = b
+                 :bar -> :bar = b
+               end
+             ) == {:ok, :dynamic}
     end
 
     test "fn" do
+      assert quoted_expr(fn :foo = b -> :foo = b end) == {:ok, :dynamic}
+
       assert {:error, _} = quoted_expr(fn :foo = b -> :bar = b end)
+    end
 
-      assert quoted_expr(fn :foo = b -> :foo = b end) == {:ok, :dynamic}
+    test "with" do
+      assert quoted_expr(
+               [a, b],
+               with(
+                 :foo <- a,
+                 :bar <- b,
+                 c = :baz,
+                 do: c
+               )
+             ) == {:ok, :dynamic}
+    end
+
+    test "for" do
+      assert quoted_expr(
+               [list],
+               for(
+                 foo <- list,
+                 is_integer(foo),
+                 do: foo == 123
+               )
+             ) == {:ok, :dynamic}
```

Comment:
This should say it returns a list of dynamic. Or we don't have return types for comprehensions yet? We also most likely need a test for something like this:

```
     assert quoted_expr(
               [list, bar],
               for(
                 foo <- list,
                 is_integer(bar),
                 do: foo == 123
               )
               bar
             ) == {:ok, :dynamic}
```

Reply:
We don't have return types for comprehensions yet because it requires analyzing the various options it takes and understanding the `Collectable` protocol.

Reply:
It is not part of this PR but the return type of `for` should be easily determined by the :into option. If it is missing or a list at compile time, then it is a list. If a literal binary, then a binary. If a literal map, a map. Otherwise it is of the same type as the expression given to “into” (and that type must be a collectable - but that’s a separate story).

Reply:
Should we hardcode the checks for lists, binaries, and maps for now then?

Reply:
Actually, I just realized the rule is that it is of the same type of whatever is in `into`. :D the default being a list. But hard coding for now is fine.